static/app/utils/formatters.tsx

```diff
@@ -148,70 +148,63 @@ export function getDuration(
-
-    const divideBy = (time: number) => {
-      return {
-        quotient: msValue < 0 ? Math.ceil(msValue / time) : Math.floor(msValue / time),
-        remainder: msValue % time,
-      };
-    };
-
-    if (value >= WEEK) {
-      const {quotient, remainder} = divideBy(WEEK);
-      const suffix = abbr ? t('wk') : ` ${tn('week', 'weeks', quotient)}`;
-
-      return `${quotient}${suffix} ${convertDuration(remainder / 1000, abbr)}`;
-    }
-    if (value >= DAY) {
-      const {quotient, remainder} = divideBy(DAY);
-      const suffix = abbr ? t('d') : ` ${tn('day', 'days', quotient)}`;
-
-      return `${quotient}${suffix} ${convertDuration(remainder / 1000, abbr)}`;
-    }
-    if (value >= HOUR) {
-      const {quotient, remainder} = divideBy(HOUR);
-      const suffix = abbr ? t('hr') : ` ${tn('hour', 'hours', quotient)}`;
-
-      return `${quotient}${suffix} ${convertDuration(remainder / 1000, abbr)}`;
-    }
-    if (value >= MINUTE) {
-      const {quotient, remainder} = divideBy(MINUTE);
-      const suffix = abbr ? t('min') : ` ${tn('minute', 'minutes', quotient)}`;
-
-      return `${quotient}${suffix} ${convertDuration(remainder / 1000, abbr)}`;
+export function getExactDuration(
+  seconds: number,
+  abbreviation: boolean = false,
+  minDuration = 'milliseconds'
+) {
+  const operation = seconds < 0 ? Math.ceil : Math.floor;
+  const levels = [
+    [operation(seconds / 604800), abbreviation ? 'wk' : ' weeks'],
+    [operation((seconds % 604800) / 86400), abbreviation ? 'd' : ' days'],
+    [operation(((seconds % 604800) % 86400) / 3600), abbreviation ? 'hr' : ' hours'],
+    [
+      operation((((seconds % 604800) % 86400) % 3600) / 60),
+      abbreviation ? 'min' : ' minutes',
+    ],
+    [
+      operation((((seconds % 604800) % 86400) % 3600) % 60),
+      abbreviation ? 's' : ' seconds',
+    ],
+    [
+      operation((((((seconds * 1000) % 604800000) % 86400000) % 3600000) % 60000) % 1000),
+      abbreviation ? 'ms' : ' milliseconds',
+    ],
+  ];
+  let returntext = '';
+
+  for (let i = 0, max = levels.length; i < max; i++) {
+    if (
+      (i === max - 1 || minDuration === (levels[i][1] as string).trim()) &&
+      !returntext &&
+      !levels[i][0]
+    ) {
+      returntext = '0' + levels[i][1];
+      break;
     }
-    if (value >= SECOND) {
-      const {quotient, remainder} = divideBy(SECOND);
-      const suffix = abbr ? t('s') : ` ${tn('second', 'seconds', quotient)}`;
-
-      return `${quotient}${suffix} ${convertDuration(remainder / 1000, abbr)}`;
+    if (levels[i][0] === 0) {
+      continue;
     }
-
-    if (value === 0) {
-      return '';
+    returntext +=
+      ' ' +
+      levels[i][0] +
+      (!abbreviation && Math.abs(levels[i][0] as number) === 1
+        ? (levels[i][1] as string).substring(0, (levels[i][1] as string).length - 1) // strip the 's' from the end if its singular
+        : levels[i][1]);
+    if (minDuration === (levels[i][1] as string).trim()) {
```

Comment:
Why does this need so much casting?

Reply:
I think `levels` is typed as `Array<[val: number, unit: string]>` it would be ok

Reply:
good call
I typed it out 👍